store/copr/batch_coprocessor.go

```diff
@@ -102,46 +103,91 @@ func (rs *batchCopResponse) RespTime() time.Duration {
 // 2. for the remaining regions:
 //    if there is only 1 available store, then put the region to the related store
 //    otherwise, use a greedy algorithm to put it into the store with highest weight
-func balanceBatchCopTask(originalTasks []*batchCopTask) []*batchCopTask {
+func balanceBatchCopTask(ctx context.Context, kvStore *kvStore, originalTasks []*batchCopTask, isMPP bool) []*batchCopTask {
 	if len(originalTasks) <= 1 {
 		return originalTasks
 	}
+	cache := kvStore.GetRegionCache()
 	storeTaskMap := make(map[uint64]*batchCopTask)
+	// storeCandidateRegionMap stores all the possible store->region map. Its content is
+	// store id -> region signature -> region info. We can see it as store id -> region lists.
 	storeCandidateRegionMap := make(map[uint64]map[string]RegionInfo)
 	totalRegionCandidateNum := 0
 	totalRemainingRegionNum := 0
 
-	for _, task := range originalTasks {
-		taskStoreID := task.regionInfos[0].AllStores[0]
-		batchTask := &batchCopTask{
-			storeAddr:   task.storeAddr,
-			cmdType:     task.cmdType,
-			ctx:         task.ctx,
-			regionInfos: []RegionInfo{task.regionInfos[0]},
+	if !isMPP {
+		for _, task := range originalTasks {
+			taskStoreID := task.regionInfos[0].AllStores[0]
+			batchTask := &batchCopTask{
+				storeAddr:   task.storeAddr,
+				cmdType:     task.cmdType,
+				ctx:         task.ctx,
+				regionInfos: []RegionInfo{task.regionInfos[0]},
+			}
+			storeTaskMap[taskStoreID] = batchTask
+		}
+	} else {
+
+		// decide the available stores
+		stores := cache.RegionCache.GetTiFlashStores()
+		var wg sync.WaitGroup
+		var mu sync.Mutex
+		wg.Add(len(stores))
+		for i := range stores {
+			go func(idx int) {
+				defer wg.Done()
+				s := stores[idx]
+				aliveReq := tikvrpc.NewRequest(tikvrpc.CmdMPPAlive, &mpp.IsAliveRequest{}, kvrpcpb.Context{})
+				aliveReq.StoreTp = tikvrpc.TiFlash
+				alive := false
+				resp, err := kvStore.GetTiKVClient().SendRequest(ctx, s.GetAddr(), aliveReq, tikv.ReadTimeoutMedium)
```

Comment:
MPP queries doesn't care about the single-digit milliseconds latency?

Reply:
Yeah, MPP queries always run for more than a second.